src/monodroid/jni/monodroid-glue.c

```diff
@@ -704,55 +737,192 @@ typedef void* (*mono_mkbundle_init_ptr) (void (*)(const MonoBundledAssembly **),
+		log_info (LOG_ASSEMBLY, "Shared library '%s' not found", path);
+		return NULL;
 	}
 
-	mono_mkbundle_initialize_mono_api = dlsym (libapp, "initialize_mono_api");
-	if (!mono_mkbundle_initialize_mono_api)
-		log_error (LOG_BUNDLE, "Missing initialize_mono_api in the application");
+	void *handle = dlopen (path, dl_flags);
+	if (handle == NULL)
+		log_info (LOG_ASSEMBLY, "Failed to load shared library '%s'. %s", path, dlerror ());
+	return handle;
+}
 
-	mono_mkbundle_init = dlsym (libapp, "mono_mkbundle_init");
-	if (!mono_mkbundle_init)
-		log_error (LOG_BUNDLE, "Missing mono_mkbundle_init in the application");
-	log_info (LOG_BUNDLE, "Bundled app loaded: %s", libappso);
+static void*
+load_dso_from_app_lib_dirs (const char *name, int dl_flags)
+{
+	if (name == NULL)
+		return NULL;
+
+	mono_bool needs_free = FALSE;
+	char *full_path = NULL;
+	for (int i = 0; i < app_lib_directories_size; i++) {
+		full_path = get_full_dso_path (app_lib_directories [i], name, &needs_free);
+		void *handle = load_dso (full_path, dl_flags, FALSE);
+		if (needs_free)
+			free (full_path);
+		if (handle != NULL)
+			return handle;
+	}
+
+	return NULL;
+}
+
+static void*
+load_dso_from_override_dirs (const char *name, int dl_flags)
+{
+	if (name == NULL)
+		return NULL;
+
+	for (int i = 0; i < MAX_OVERRIDES; i++) {
+		if (override_dirs [i] == NULL)
+			continue;
+
+		mono_bool needs_free = FALSE;
+		char *full_name = get_full_dso_path (override_dirs [i], name, &needs_free);
+		void *handle = load_dso (full_name, dl_flags, FALSE);
+		if (needs_free)
+			free (full_name);
+		if (handle != NULL)
+			return handle;
+	}
+
+	return NULL;
 }
 
 static char*
-get_bundled_app (JNIEnv *env, jstring dir)
+get_existing_dso_path_on_disk (const char *base_dir, const char *dso_name, mono_bool *needs_free)
 {
-	const char *v;
-	char *libapp;
+	mono_bool _dummy_needs_free = FALSE;
+	if (needs_free == NULL)
+		needs_free = &_dummy_needs_free;
 
-#ifndef RELEASE
-	libapp = path_combine (override_dirs [0], "libmonodroid_bundle_app.so");
+	*needs_free = FALSE;
+	char *dso_path = get_full_dso_path (base_dir, dso_name, needs_free);
+	if (file_exists (dso_path))
+		return dso_path;
 
-	if (file_exists (libapp))
-		return libapp;
+	*needs_free = FALSE;
+	free (dso_path);
+	return NULL;
+}
 
-	free (libapp);
-#endif
+static void
+__dso_alloc_cleanup (char **dso_path, mono_bool *needs_free)
+{
+	if (dso_path != NULL) {
+		if (needs_free && *needs_free)
+			free (*dso_path);
+		*dso_path = NULL;
+	}
 
-	if (dir) {
-		v = (*env)->GetStringUTFChars (env, dir, NULL);
-		if (v) {
-			libapp = path_combine (v, "libmonodroid_bundle_app.so");
-			(*env)->ReleaseStringUTFChars (env, dir, v);
-			if (file_exists (libapp))
-				return libapp;
-		}
+	if (needs_free)
+		*needs_free = FALSE;
+}
+
+static char*
+get_full_dso_path_on_disk (const char *dso_name, mono_bool *needs_free)
+{
+	mono_bool _dummy_needs_free = FALSE;
+	if (needs_free == NULL)
```

Comment:
Given that ~all of these functions are `static` and can thus be audited, perhaps we should just `assert(needs_free)` or let the process crash if we access `*needs_free`?  Is there a codepath where we *deliberately* pass `NULL` to `needs_free`?  'cause I'm seeing lots of handling of a NULL value, but every codepath so far has always provided the value, so we have lots of conditional code which should never need to happen.

Reply:
We're not passing `NULL` anywhere, but I don't like writing unguarded code. `assert` is a good idea.

Reply:
I saw the use of `assert(needs_free)` elsewhere.  Why not here?